### ISSUE TITLE
Performance regression in OCaml 4.08

### ISSUE THREAD

**Issue by ejgallego** [ocaml/ocaml#9326]
Dear OCaml devs,

@RalfJung has pointed out a non-trivial performance regression in Coq when moving from OCaml 4.07 to OCaml 4.08 https://github.com/coq/coq/issues/11652

See some numbers [here](https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/861/console), for example CompCert does compile more than 7% faster in 4.07

There are a few candidates in the changelog for the slowdown, I think we will try reverting some PRs and seeing what happens with a reduced file. Any ideas? [For example #2144 could be relevant? I doubt it tho]

Benchs:
- 4.07.1 vs 4.10.0 : https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/865/
- 4.07.1 vs 4.09.0: https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/864/
- 4.08.1 vs 4.07.1: https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/861/
- 4.09.0 vs 4.08.1: https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/863/

**Comment by ejgallego** [comment]
[by the way, do you know folks where are the OCaml performance test stored?]

**Comment by gasche** [comment]
There are no performance tests in the compiler distribution (some parts of the testsuite, typically `misc-kb`, also serve as microbenchmarks, but we do not monitor their performance counters). We have benchmarks in [operf-micro](https://github.com/OCamlPro/operf-micro) and [operf-macro](https://github.com/OCamlPro/operf-macro), and two benchmark-result-monitoring websites (<http://bench.flambda.ocamlpro.com/>, <http://bench2.ocamllabs.io:8083/>). Those websites are used by specific groups (flambda developers, multicore developers), but as far as I remember they do not typically report performance regressions in the standard OCaml trunk.

**Comment by ejgallego** [comment]
Thanks @gasche , I was wondering if some of the people measuring perf had found the same regression as us.

I was not able to get a meaninful result from the ocamlpro site; I knew about the ocamllabs site but it seems limited to 4.06+multicore, right?

**Comment by gasche** [comment]
Regarding where regressions may be: many items in the changelog could lead to performance differences, and it's fairly hard to tell. One first thing to look at may be that 4.08 is the first release to use an autoconf-based configuration (before OCaml had a manually-written configure script). Because the various feature detection logic has been rewritten, it is possible that a stock 4.08 compiler gets configured differently from a stock 4.07 compiler in a way that impacts performances.

@ejgallego I don't remember reports of similar regressions from industrial OCaml users (who typically do run performance tests on their own programs when they switch to a new OCaml version).

**Comment by ejgallego** [comment]
We can reproduce with a standard `opam` setup; in that case I cannot witness any difference in configuration that should matter, but indeed we need to check this more carefully.

> @ejgallego I don't remember reports of similar regressions from industrial OCaml users (who typically do run performance tests on their own programs when they switch to a new OCaml version).

In fact I was wondering about this precise point, how come nobody else had noticed yet. I guess it could be something very specific related to Coq; or maybe some compute-intensive users are still in 4.07 ?

**Comment by ejgallego** [comment]
Also, I remember seeing quite degraded performance with 4.08+flambda some time back; I didn't research a lot as I had no time back then.

**Comment by gasche** [comment]
I believe that at least Lexifi and Jane Street have migrated to 4.08 and then 4.09.

One candidate for a performance regression for Coq could be https://github.com/ocaml/ocaml/pull/1895 : it fixes the capture of callstacks/backtraces in new threads, which would previously only capture the last stack frame. If Coq captures callstacks/backtraces and does computations in threads, it could be hit by a performance regression coming from the callstacks suddenly being much longer.

There are many other PRs that could be performance regressions if we are unlucky. For example #2060 and #2070 might result in deoptimization if there is a mistake in them. #2083 is a bugfix that disables some unsafe unboxing in presence of GADT; if a single hot function from Coq is affected, it could give a visible performance degradation.

Seeing a performance profile of a Coq run under both versions might help in narrowing the issue down?

**Comment by ejgallego** [comment]
Thanks a lot for the pointers @gasche ; indeed the preliminary plan could look like:
- wait for the 4.09 results
- narrow down from full development to a minimal Coq `.v` that allow us to reproduce the 30% slowdown Ralf experienced.
- get perf profiles
- try reverting some of the patches you propose

IMHO it is still hard to tell what the cause could be; all of the options you propose are for now under the table.

**Comment by gasche** [comment]
Can you confirm that the numbers given are for x86_64 machines, in particular not 32bit machines? (For example #1683 affects marshalling of Custom values, and I think Coq uses Custom blocks to represent VM's uint63 on 32-bits machines.)

**Comment by ejgallego** [comment]
For now x86_64 , also the slowdowns don't seem to be tied to developments making heavy use of the VM.

**Comment by nojb** [comment]
We suspect multiple sources of slowdowns around the 4.08 release, but unfortunately they haven't yet been tracked down precisely.

Have you seen #8776 ?

**Comment by ejgallego** [comment]
Thanks @nojb ; note that in this case we are not talking about the compiler itself, but the code _produced_ by the compiler is in some cases 30% slower.

**Comment by ejgallego** [comment]
Of course would the compiler itself be affected by same Coq bug that would explain why there is an overall slowdown.

I wonder if systematic benchs should be introduced at the PR level; in Coq we are limited by the amount of available hardware (so we only test selected PRs) but the plan is to do so.

**Comment by gasche** [comment]
Well, PR-level performance monitoring is of course a great idea, but it is also a lot of work, and so far no one has been willing to do this consistently for the OCaml trunk.

(#8776 is a compile-time regression due to a change in type-checking behavior, it should not influence programs that do not call the type-checker at runtime.)

**Comment by xavierleroy** [comment]
Is the slowdown observed with the flambda optimizer or without?  

Without flambda and using KB and a few other old tests as benchmarks, I'm not seeing significant performance differences between 4.07.1, 4.08.1, and 4.09.0.

Will try flambda next.

**Comment by gasche** [comment]
I believe that Emilio's [numbers](https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/861/consoleText) come from the standard ocaml-base-compiler-{4.07.1,4.08.1} switches, with default configuration options, so in particular without flambda. I reviewed the opam files to check if there were configuration differences, and did not find much. (4.07.1 uses `-with-debug-runtime`, and I don't know whether it is enabled by default in 4.08, but this should not change anything).

@ejgallego: your 4.08/4.09 benchmark failed because it appears to use an invalid switch name: there is no 4.09.1 release yet.

**Comment by RalfJung** [comment]
> narrow down from full development to a minimal Coq .v that allow us to reproduce the 30% slowdown Ralf experienced.

In case that's useful, [this is the file with the biggest slowdown](https://gitlab.mpi-sws.org/iris/iris/blob/master/theories/algebra/namespace_map.v), of 28%. There are [some more that are almost as bad](https://coq-speed.mpi-sws.org/d/1QE_dqjiz/coq-compare?orgId=1&var-project=iris&var-branch1=master&var-commit1=7ddea37a842878fa5a9eb53130e0f867aced2655&var-config1=build-coq.8.11.0&var-branch2=master&var-commit2=4815e68e8bad53eb3fb832d39c282ea8ac36a676&var-config2=build-coq.8.11.0&var-group=(.*)&var-metric=instructions), but they have even longer dependency chains.

**Comment by ejgallego** [comment]
> Is the slowdown observed with the flambda optimizer or without?

The numbers on the Coq bench are without flambda, I am not sure what @RalfJung is using in their CI but given that he spoke about the "Debian" version that would make it flambda [for Iris we use flambda in Coq's CI as the improvement is noticeable.

I do remember seeing an slowdown in 4.08.0+flambda too, I will run some benchs with flambda once the current one finish.

> @ejgallego: your 4.08/4.09 benchmark failed because it appears to use an invalid switch name: there is no 4.09.1 release yet.

@gasche indeed I got confused by opam having a 4.09.1 ocaml package; link updated; I've also added a 4.07.1 vs 4.09.0 bench.

**Comment by RalfJung** [comment]
>  I am not sure what @RalfJung is using in their CI

We are using the `ocaml-base-compiler.4.07.1` and `ocaml-base-compiler.4.08.1` as installed via opam, respectively. I used the Debian version number merely as an indicator that upgrading from 4.07 to 4.08 is reasonable.

So, no flambda.

**Comment by gasche** [comment]
Note: Emilio's benchmarks are configured so that "NEW" is the older OCaml compiler, and "OLD" is the newer OCaml compiler. So -7% on the [first benchmarks: 4.07.1 vs 4.08.1](https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/861/consoleText) means that Compcert builds 7% *faster* on 4.07.1 than 4.08.1, and +2% on the [second benchmarks: 4.08.1 vs 4.09.0](https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/863/consoleText) indicates that Compcert builds 2% *slower* on 4.08.1 than 4.09.0.
(This choice is very confusing!)

**Comment by ejgallego** [comment]
> (This choice is very confusing!)

Sorry for that, it was not intended; I got confused in the UI.

**Comment by ejgallego** [comment]
Benchs updated; compiling ssreflect takes the longest hit [around 10% on average] so we will try work a reduced test case from that.

**Comment by ejgallego** [comment]
> Without flambda and using KB and a few other old tests as benchmarks, I'm not seeing significant performance differences between 4.07.1, 4.08.1, and 4.09.0.

@xavierleroy for now the easiest way to reproduce is to compile ssreflect ; we are working on isolating a reduced case from that.

**Comment by ejgallego** [comment]
Numbers from the 4.07.1 vs 4.10.0 build:

```
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-bignums │   65.88   63.21  +4.22 % │      182290289689      175181022221  +4.06 % │      259976427290      244726082903  +6.23 % │     464956     483176  -3.77 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fiat-core │  107.03  102.69  +4.23 % │      302449462088      291188243030  +3.87 % │      420512910625      388954324356  +8.11 % │     482324     497936  -3.14 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-solvable │  185.49  176.74  +4.95 % │      515966048509      492177773809  +4.83 % │      750818902765      697493194786  +7.65 % │     754736     793260  -4.86 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                    coq-corn │ 1507.07 1434.43  +5.06 % │     4200011413224     4005363120587  +4.86 % │     6662181658471     6158159851087  +8.18 % │     793812     848800  -6.48 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-math-classes │  216.76  206.20  +5.12 % │      602524660136      576835221965  +4.45 % │      827481589713      770360634497  +7.41 % │     498344     518240  -3.84 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                  coq-sf-plf │   46.44   44.12  +5.26 % │      128883308156      122875098405  +4.89 % │      183445777749      168750428174  +8.71 % │     473424     494872  -4.33 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-compcert │  718.74  681.33  +5.49 % │     2004646793353     1898626696774  +5.58 % │     2896758083193     2648267852440  +9.38 % │    1035420    1037784  -0.23 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-color │  695.52  656.16  +6.00 % │     1939340221290     1832028028189  +5.86 % │     2574994314407     2319850505090 +11.00 % │    1127676    1140584  -1.13 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-fingroup │   52.42   49.44  +6.03 % │      145313028762      137436705962  +5.73 % │      207668212323      190242453986  +9.16 % │     548796     557196  -1.51 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-ssreflect │   53.72   49.58  +8.35 % │      147220475940      136074472580  +8.19 % │      200823608869      176490732956 +13.79 % │     554532     568724  -2.50 % │    0    0  +nan % │
└─────────────────────────────┴──────────────────────────┴──────────────────────────────────────────────┴──────────────────────────────────────────────┴────────────────────────────────┴───────────────────┘
```

Note that this are better than just 4.07.1 to 4.08.1 ; there is also a big improvement in a few developments in 4.10 but not in 4.08.1 so IMO that's orthogonal to the regression at question here.

**Comment by RalfJung** [comment]
Are some entries missing from that table? lambdarust/iris should usually be in there.

**Comment by gasche** [comment]
There is this line in the 4.08.1 vs 4.07.1 results:

``` 
│        coq-lambda-rust │ 1544.30 1594.67  -3.16 % │     4305486348742     4441777015853  -3.07 % │     6208685582582     6423540309827  -3.34 % │    1135604    1129516  +0.54 % │     0     0  +nan % │
```

This suggests that 4.07.1 is 3.16% faster than 4.08.1 at checking coq-lambda-rust.

I have looked at the numbers in general and I think that they point out at a clear regression between 4.07.1 and 4.08.1 (with some independent improvements in further versions). There is not much that I could tell from these numbers, and I stopped looking at them.

To go further I would personally need one of:
- perf/gprof profiles for the Coq runs that take a performance hit
- result on whether some of the PRs that were pointed above may be guilty
- or in general a bisection of the performance between 4.07.1 and 4.08.1 (but I suspect that this would be a massive amount of work)

**Comment by lpw25** [comment]
I would add to Gabriel's list:

- Some indication of how noisy these numbers are for other changes to coq. These numbers are within the bounds of what can be caused by comparing a lucky code layout with an unlucky one, and it would be good to know that other changes which affect the code layout don't produce similar changes in performance.

**Comment by ppedrot** [comment]
@lpw25 usually the noise range on our bench is around 1%.

**Comment by xavierleroy** [comment]
I was able to reproduce the slowdown by running Coq on CompCert files.  It looks like the major GC is working harder in 4.08.1 than in 4.07.1, in the context of Coq at least.  @damiendoligez is looking into it.

**Comment by ejgallego** [comment]
Dear all, some new results w.r.t. 4.11:

- https://github.com/coq/coq/pull/13038#issuecomment-694239621
- https://github.com/coq/coq/pull/13040#issuecomment-695335651

Even if best_fit seems to improve a some cases, some users seem to be still hit (4.07.1 vs 4.11.1 + best_fit) :
```
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│        coq-mathcomp-algebra │  141.93  140.34  +1.13 % │      393248098935      390141024521  +0.80 % │      513136676800      503715278515 +1.87 % │     574280     565036  +1.64 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│              coq-coquelicot │   70.84   69.87  +1.39 % │      194607296865      191135104045  +1.82 % │      254996589554      245303622507 +3.95 % │     599300     587168  +2.07 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-compcert │  678.71  665.87  +1.93 % │     1889485385741     1856340836411  +1.79 % │     2649341650215     2555954385433 +3.65 % │    1124164    1131392  -0.64 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-math-classes │  193.92  190.08  +2.02 % │      540195760748      530267944670  +1.87 % │      719549432958      700334701606 +2.74 % │     498856     504748  -1.17 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-color │  633.18  617.46  +2.55 % │     1766537049254     1724694289532  +2.43 % │     2292326187434     2190545821844 +4.65 % │    1638820    1455240 +12.62 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fiat-core │  110.22  105.95  +4.03 % │      312971015270      301974056873  +3.64 % │      426957676501      401523424965 +6.33 % │     484804     487364  -0.53 % │    0    0  +nan % │
├─────────────────────────────┼──────────────────────────┼──────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
```

**Comment by gasche** [comment]
As far as I know, we still have no idea of which change precisely in 4.08+dev is introducing the regression. What should we do to work on this?

In https://github.com/ocaml/ocaml/issues/9326#issuecomment-590000294 I mentioned a handful of PRs that might be involved in the regression. Maybe I could try to provide experimental variants of 4.08 with those PRs *reverted*, and then you ( @ejgallego ) would run the benchmarks (or just a couple benchmark that suffice the regression) on them?

**Comment by ejgallego** [comment]
I'm afraid I don't have the cycles to run the bisect right now, sorry.

Something that would make this much easier would be to allow to compose the compiler build with OCaml libraries, as we support for Coq when using dune.

**Comment by ejgallego** [comment]
> In [#9326 (comment)](https://github.com/ocaml/ocaml/issues/9326#issuecomment-590000294) I mentioned a handful of PRs that might be involved in the regression. Maybe I could try to provide experimental variants of 4.08 with those PRs _reverted_, and then you ( @ejgallego ) would run the benchmarks (or just a couple benchmark that suffice the regression) on them?

That would help, thanks, however our bench system has changed and it is much harder now to try new switches / package subsets :S

**Comment by stedolan** [comment]
I dug into this a bit, and I think I have an explanation. The cause is a bug in `Gc.set` that incorrectly changes some GC settings.

I could reproduce the issue by comparing builds of mathcomp-1.12 using Coq 8.12.2 on OCaml 4.07.1 and 4.08.1. Many of the coqc invocations are ~10% slower on 4.08.1. It's not a code layout issue (4.08.1 is actually executing ~10% more instructions), and the extra time is spent in GC. I arbitrarily picked one file (`ssreflect/choice.v`) that saw a ~10% slowdown, with the full command being:
```
coqc -q -w -projection-no-head-constant -w -redundant-canonical-projection -w -notation-overridden \
  -w +duplicate-clear -w -ambiguous-paths -w +non-primitive-record -w +undeclared-scope \
  -I . -R . mathcomp ssreflect/choice.v
```

Coq uses [non-default GC settings for speed](https://github.com/coq/coq/blob/master/toplevel/coqtop.ml#L142-L156), corresponding to an `OCAMLRUNPARAM` setting of `s=33554432,o=120`. These settings are applied at startup using `Gc.set`, unless the environment variable `OCAMLRUNPARAM` is explicitly set.

Weirdly, the performance difference goes away when the environment variable is set:

|                  | 4.07.1 | 4.08.1 |
|------------------|--------|--------|
| no OCAMLRUNPARAM | 0.57 s | 0.64 s |
| s=33554432,o=120 | 0.57 s | 0.58 s |
| s=262144,o=80    | 0.81 s | 0.83 s |

The first two rows should be identical: the second row is explicitly setting what the first row uses as a default.

The issue is that Coq does `Gc.set { Gc.get () with ... }`, and `Gc.set` has a bug: integer untagging is not correctly performed when reading the `custom_major_ratio`, `custom_minor_ratio` and `custom_minor_max_bsz` fields that were introduced in 4.08.1. This is visible when using the toplevel in verbose GC mode:
```
$ OCAMLRUNPARAM=v=0x20 ocaml
Initial minor heap size: 256k words
Initial major heap size: 992k bytes
Initial space overhead: 80%
Initial max overhead: 500%
Initial heap increment: 15%
Initial allocation policy: 0
Initial smoothing window: 1
        OCaml version 4.11.1

# Gc.set (Gc.get ());;
New custom major ratio: 89%
New custom minor ratio: 201%
New custom minor size limit: 16385%
- : unit = ()
# Gc.set (Gc.get ());;
New custom major ratio: 179%
New custom minor ratio: 403%
New custom minor size limit: 32771%
- : unit = ()
# Gc.set (Gc.get ());;
New custom major ratio: 359%
New custom minor ratio: 807%
New custom minor size limit: 65543%
- : unit = ()
# Gc.set (Gc.get ());;
New custom major ratio: 719%
New custom minor ratio: 1615%
New custom minor size limit: 131087%
- : unit = ()
```

This means that every call to `Gc.set` doubles the GC parameters used to control collection of custom blocks. This will have a strong effect on programs that read many files using the stdlib's IO functions, as each channel is a 64 KB custom block.

**Comment by gasche** [comment]
(I arrived just in time to reserve the last cool smiley for myself.) Thanks for the fascinating find!

It is surprising that other users have not caught the issue before. It sounds like explicit GC settings are not that common, or people don't do that much differential performance testing against compiler versions. There should be a small-but-noticeable performance regression in all programs that use explicit Gc settings (in the code) and also custom blocks.

The bug is very easy to fix. If I understand correctly, fixing it may improve the performance of Coq (and other programs) on all >=4.08 versions.

**Comment by xavierleroy** [comment]
Great detective work!  @damiendoligez and I did observe that the major GC was working harder in 4.08 than in 4.07 on those runs of Coq, but the root cause was unknown.  In retrospect, it also explains why some runs of Coq are affected more than others: some runs read more files than others.

**Comment by xavierleroy** [comment]
Obvious fix + regression test in #10125.

**Comment by JasonGross** [comment]
For other archeologists like me:

> Coq uses [non-default GC settings for speed](https://github.com/coq/coq/blob/master/toplevel/coqtop.ml#L142-L156), corresponding to an `OCAMLRUNPARAM` setting of `s=33554432,o=120`. These settings are applied at startup using `Gc.set`, unless the environment variable `OCAMLRUNPARAM` is explicitly set.

The link now points to the wrong code, the correct perma-link is [non-default GC settings for speed](https://github.com/coq/coq/blob/8c7457e18de2fb5be89f22c76ac59541345d1d5c/toplevel/coqtop.ml#L142-L156), and the current permalinked location as of the writing of this comment is [here](https://github.com/coq/coq/blob/f5743b193e5179175b97142c4c8c9aa6a76def29/sysinit/coqinit.ml#L11-L31)
